h2/src/main/org/h2/command/dml/Select.java

```diff
@@ -847,9 +854,11 @@ public void prepare() {
             sort = prepareOrder(orderList, expressions.size());
             orderList = null;
         }
-        for (int i = 0; i < expressions.size(); i++) {
-            Expression e = expressions.get(i);
-            expressions.set(i, e.optimize(session));
+        if (!isExistsSubquery) {
+            for (int i = 0; i < expressions.size(); i++) {
+                Expression e = expressions.get(i);
+                expressions.set(i, e.optimize(session));
+            }
```

Comment:
> I would rather things like "1/0" would fail at prepare time

It looks weird indeed, but try it in any other database. E.g. PostgreSQL, Oracle, SQL Server, MySQL all skip evaluating the `SELECT` clause inside of an `EXISTS()` predicate.

The reason why I chose `1 / 0` in the integration tests is because this is a very easy way to verify the current implementation. But I could also change the PR to go forward with the optimization step, and run that division against an actual table...

> And yeah, I am nervous about this codebase

It's a good attitude :)